File: Behavioral/Mediator/Cpp/ChatUser.cpp
// ChatUser.cpp

#include "ChatUser.h"
#include <iostream>

ChatUser::ChatUser(Mediator* mediator, std::string name) : User(mediator, std::move(name)) {}

void ChatUser::sendMessage(const std::string& message) {
  std::cout << name << " sends: " << message << std::endl;
  mediator->sendMessage(message, this);
}

void ChatUser::receiveMessage(const std::string& message) {
  std::cout << name << " receives: " << message << std::endl;
}

File: Behavioral/Mediator/CSharp/ChatUser.cs
// ChatUser.cs

using System;

public class ChatUser : User {
  public ChatUser(IMediator mediator, string name) : base(mediator, name) {}

  public override void SendMessage(string message) {
    Console.WriteLine($"{name} sends: {message}");
    mediator.SendMessage(message, this);
  }

  public override void ReceiveMessage(string message) {
    Console.WriteLine($"{name} receives: {message}");
  }
}
